### ISSUE TITLE
Form doesn't wait for pending validations before submitting.

### ISSUE THREAD

**Issue by nat-n** [final-form/react-final-form#177]
I've found something that looks like a bug.

I don't think it should be possible to submit a form until validations have passed, however if validation is in progress a form still submits as if the validation had passed..

The following example is just like the Async validation example but with a longer delay inside the validator. Filling in the form with an invalid name and immediately submitting allows the form to be submitted with invalid content.

https://codesandbox.io/s/8n1n7zy1rj

I believe form submission should wait for validation to complete. Furthermore ensuring appropriate timeouts so that a form isn't locked by a never ending validation should be the concern of the validation function.

**Comment by declanelcocks** [comment]
@nat-n You can change your submit button to this:

```js
<button type="submit" disabled={validating || submitting}>
```

I think we can assume that `submitting` is setting the precedent for how these kind of properties should be consumed. The library itself only takes care of setting `submitting` to `true` or `false`, and it's up to us to decide how the UI behaves in each situation.

In all of the examples, the `<button />` is manually disabled based on the value of `submitting`, so I don't think `validating` would be any different. 

Given the trend of React libraries exposing an API to you rather than a UI, I think it's safe to assume that it's on us, the users, to decide how we use these properties.

**Comment by nat-n** [comment]
@declanelcocks 
I see how the API makes this issue manageable, but I'm not convinced that it's not an issue worth addressing at the root, or that doing so would present a compromise. The fact that the one officially documented example of Async-Validation actually contains a subtle version of this bug is telling.

Maybe this is a discussion for the final-form repo? But the key difference with submission here –
 for which you might want to enable multiple parallel submits and handle the implications of this downstream – is that the partially enforced implication of validation is that submission is not allowed when the form is not in a valid state.

So I think a more consistent and less surprising API is possible here, that doesn't take anything from the user, but does remove a gotcha, and make it more natural to implement better UX's; such as that if that user attempts to submit while a validation is still running, the submit execution is delayed pending the validation. This seems obviously better than enable/disabling the submit button for reasons that the user might not be aware of and shouldn't have to care about.

**Comment by declanelcocks** [comment]
True, I didn't think about the fact that you're technically submitting an invalid form, which is one of the things the library is trying to prevent. Maybe it's fine to just exit out of the submit if any field is `invalid || validating`.

**Comment by Zn4rK** [comment]
This bug has been patched in ```final-form@4.2.1``` (via https://github.com/final-form/final-form/issues/82, https://github.com/final-form/final-form/pull/83).

If you update the dependencies in the codesandbox, it works as intended:
https://codesandbox.io/s/j4noyqymry
